internal/configs/provider_validation.go

```diff
@@ -95,6 +142,44 @@ func validateProviderConfigs(parentCall *ModuleCall, cfg *Config, noProviderConf
 		}
 	}
 
+	checkImpliedProviderNames := func(resourceConfigs map[string]*Resource) {
+		// Now that we have all the provider configs and requirements validated,
+		// check for any resources which use an implied localname which doesn't
+		// match that of required_providers
+		for _, r := range resourceConfigs {
+			// We're looking for resources with no specific provider reference
+			if r.ProviderConfigRef != nil {
+				continue
+			}
+
+			localName := r.Addr().ImpliedProvider()
+			if _, ok := localNames[localName]; ok {
+				// OK, this was listed directly in the required_providers
+				continue
+			}
+
+			defAddr := addrs.NewDefaultProvider(localName)
```

Comment:
I think this could instead be a call to [`Module.ImpliedProviderForUnqualifiedType`](https://github.com/hashicorp/terraform/blob/f266d1ee82d1fa4d882c146cc131fec4bef753cf/internal/configs/module.go#L573-L577), which I believe aims to encapsulate the logic for "what is the fallback to use if this doesn't match an explicit provider requirement?".

That _does_ also consult the required providers for the module, and so calling it directly wouldn't allow distinguishing between a reference to an explicit declaration vs. a backward-compatibility fallback. So maybe then the better answer would be to skip over that layer and go directly to `addrs.ImpliedProviderForUnqualifiedType`, which at least deals with the special case that `terraform` means `terraform.io/builtin/terraform` rather than `hashicorp/terraform`. Or maybe we could change the signature of `Module.ImpliedProviderForUnqualifiedType` to return an additional result that signals explicitly whether this was taken from an explicit `required_providers` declaration or just a compatibility fallback, in the interests of keeping all of the logic using mostly the same codepaths. :thinking: 

Generally I'd encourage leaning on these existing methods as much as they make sense, because they are there to encapsulate all of the little oddities that arose in the process of developing the provider source scheme, and so if we skip over them and try to reimplement their logic out here it's likely that we'll miss some edge case that we've forgotten about in the meantime and make this not behave consistently with the logic elsewhere.

Reply:
I don't remember us making any special effort to treat `source = "terraform.io/builtin/terraform"` as invalid, so as far as I know you _can_ include it there if you really want to give it a local name other than `terraform`, though of course that would be a pretty bizarre thing to do.

Reply:
oh yeah, you can, not sure why I thought we disallowed that 🤔